osu.Desktop/GameStateBroadcaster.cs

```diff
@@ -0,0 +1,187 @@
+// See the LICENCE file in the repository root for full licence text.
+
+using System;
+using System.Collections.Concurrent;
+using System.Net;
+using System.Net.WebSockets;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Newtonsoft.Json;
+using osu.Framework.Allocation;
+using osu.Framework.Bindables;
+using osu.Framework.Graphics;
+using osu.Framework.Threading;
+using osu.Game.Beatmaps;
+using osu.Game.Rulesets;
+
+namespace osu.Desktop
+{
+    public class GameStateBroadcaster : Component
+    {
+        private readonly ConcurrentBag<WebSocketClient> clients = new ConcurrentBag<WebSocketClient>();
+
+        private readonly HttpListener listener = new HttpListener();
+
+        private readonly CancellationTokenSource cancellationToken = new CancellationTokenSource();
+
+        private readonly Bindable<RulesetInfo> ruleset = new Bindable<RulesetInfo>();
+
+        private readonly Bindable<WorkingBeatmap> beatmap = new Bindable<WorkingBeatmap>();
+
+        private ScheduledDelegate broadcastSchedule;
+
+        public GameStateBroadcaster()
+        {
+            listener.Prefixes.Add("http://localhost:8080/");
+            listener.Start();
+
+            var thread = new Thread(() => handleConnections().Wait())
+            {
+                Name = "GameStateBroadcaster",
+                IsBackground = true,
+            };
+
+            thread.Start();
+        }
+
+        [BackgroundDependencyLoader]
+        private void load(Bindable<RulesetInfo> ruleset, Bindable<WorkingBeatmap> beatmap)
+        {
+            this.ruleset.BindTo(ruleset);
+            this.beatmap.BindTo(beatmap);
+
+            this.ruleset.ValueChanged += _ => broadcast();
+            this.beatmap.ValueChanged += _ => broadcast();
+        }
+
+        private void broadcast()
+        {
+            broadcastSchedule?.Cancel();
+            broadcastSchedule = Scheduler.AddDelayed(() =>
+            {
+                string state = JsonConvert.SerializeObject(new GameState
+                {
+                    Ruleset = ruleset.Value,
+                    BeatmapMetadata = beatmap.Value.Metadata,
+                });
+
+                foreach (var client in clients)
+                    client.Enqueue(state);
+            }, 200);
+        }
+
+        public async Task Close()
+        {
+            foreach (var client in clients)
+                await client.Close().ConfigureAwait(false);
+        }
+
+        private async Task handleConnections()
+        {
+            while (!cancellationToken.IsCancellationRequested)
+            {
+                var context = await listener.GetContextAsync().ConfigureAwait(false);
+                var wsContext = await context.AcceptWebSocketAsync(null).ConfigureAwait(false);
+
+                var client = new WebSocketClient(wsContext.WebSocket);
+                clients.Add(client);
+
+                var thread = new Thread(() => client.Handle().Wait())
```

Comment:
All this `new Thread` stuff cannot exist. Please read up on newer .NET paradigms and follow them, as I'm 12020% sure that this is not the correct way.